rust/noosphere/src/ffi/sphere.rs

```diff
@@ -86,3 +89,115 @@ pub fn ns_sphere_join(
+/// base64-encoded CID v1 string.
+pub fn ns_sphere_version_get(
+    noosphere: &NsNoosphereContext,
+    sphere_identity: char_p::Ref<'_>,
+    error_out: Option<Out<'_, repr_c::Box<NsError>>>,
+) -> Option<char_p::Box> {
+    error_out.try_or_initialize(|| {
+        noosphere.async_runtime().block_on(async {
+            let sphere_context = noosphere
+                .inner()
+                .get_sphere_context(&Did(sphere_identity.to_str().into()))
+                .await?;
+
+            let sphere_context = sphere_context.lock().await;
+            sphere_context
+                .sphere()
+                .await?
+                .cid()
+                .to_string()
+                .try_into()
+                .map_err(|error: InvalidNulTerminator<String>| anyhow!(error).into())
+        })
+    })
+}
+
+#[ffi_export]
+/// Sync a sphere with a gateway. A gateway URL must have been configured when
+/// the [NoosphereContext] was initialized. And, the sphere must have already
+/// been created or joined by the caller so that it is locally initialized (it's
+/// okay if this was done in an earlier session). The returned string is the
+/// base64-encoded CID v1 of the latest locally-available sphere revision after
+/// the synchronization process has successfully completed.
+pub fn ns_sphere_sync(
+    noosphere: &mut NsNoosphereContext,
+    sphere_identity: char_p::Ref<'_>,
+    error_out: Option<Out<'_, repr_c::Box<NsError>>>,
+) -> Option<char_p::Box> {
+    error_out.try_or_initialize(|| {
+        let cid = noosphere.async_runtime().block_on(async {
+            let sphere_context = noosphere
+                .inner()
+                .get_sphere_context(&Did(sphere_identity.to_str().into()))
+                .await?;
+
+            let mut sphere_context = sphere_context.lock().await;
+
+            sphere_context.sync().await?;
+
+            Ok(sphere_context.sphere().await?.cid().to_string()) as Result<String, anyhow::Error>
+        })?;
+
+        Ok(cid
+            .try_into()
+            .map_err(|error: InvalidNulTerminator<String>| anyhow!(error))?)
+    })
+}
+
+#[ffi_export]
+/// Get an array of all of the slugs that changed in a given sphere since a
+/// given revision of that sphere (excluding the given revision). The revision
+/// should be provided as a base64-encoded CID v1 string. If no revision is
+/// provided, the entire history will be considered (back to and including the
+/// first revision).
+///
+/// Note that a slug change may mean the slug was added, updated or removed.
+/// Also Note also that multiple changes to the same slug will be reduced to a
+/// single entry in the array that is returned.
+pub fn ns_sphere_changed_slugs(
+    noosphere: &NsNoosphereContext,
+    sphere_identity: char_p::Ref<'_>,
+    since_cid: Option<char_p::Ref<'_>>,
+    error_out: Option<Out<'_, repr_c::Box<NsError>>>,
+) -> c_slice::Box<char_p::Box> {
+    let possible_output = error_out.try_or_initialize(|| {
```

Comment:
That's a great idea, and yes I think it probably could be made into a macro. That said, we do sometimes care about the return value so I'm not sure that the calling semantics would look all that different in the end. I'll file an issue so that we can investigate further.

#211